File: C++/Binary_Tree_Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ret;
        if (root == NULL) return ret;
        ret = binaryTreePaths(root->left);
        vector<string>tmp = binaryTreePaths(root->right);
        ret.insert(ret.end(), tmp.begin(), tmp.end());
        if (ret.size() == 0) ret.push_back(to_string(root->val));
        else {
            for (string &s : ret) {
                s = to_string(root->val) + "->" + s;
            }
        }
        return ret;
    }
};

File: Java/Binary_Tree_Paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    List<String> ret;
    public List<String> binaryTreePaths(TreeNode root) {
        ret = new LinkedList<String>();
        StringBuilder pred = new StringBuilder();
        if (root == null) return ret;
        solve(root, pred);
        return ret;
    }
    private void solve(TreeNode root, StringBuilder pred) {
        int count_added = 0;
        if (pred.length() > 0) {
            pred.append("->");
            count_added = 2;
        }
        pred.append(root.val);
        count_added += Integer.toString(root.val).length();
        if (root.left == null && root.right == null) {
            ret.add(pred.toString());
        }
        if (root.left != null) {
            solve(root.left, pred);
        }
        if (root.right != null){
            solve(root.right, pred);
        }
        pred.setLength(pred.length()-count_added);
    }
}
